Feature: Replace Makefile with MSBuild

SeeSharpDeveloper opened the pull request:
## Changes

- Added Template props and Targets to enable build, compose and setup of docker and npm with msbuild commands
- Update Readme
- Changed order of targets with most significant target at top
- moved branch - accidently stopped PR #22

## Tasks

[factro](https://cloud.factro.com/5681c8b0/?p=task&pi=d1d49ee7-216d-4cec-ab50-7287a344614d)

PR: #24

SeeSharpDeveloper committed:
ADD props and targets for makefile replacement

SeeSharpDeveloper committed:
ADD: Template Installation in MsBuild
instead of Makefile

SeeSharpDeveloper committed:
UPDATE: Readme for Project and Template

SeeSharpDeveloper committed:
FIX: add **/frontend/build/ to .gitignore

SeeSharpDeveloper committed:
UPDATE:  Readme: Stop Docker

gazer42 committed:
fix makefile

gazer42 committed:
enhance ui-tests option description

gazer42 committed:
increase template version 2.1.10 -> 2.2.0

gazer42 committed:
extend readme for template developer

SeeSharpDeveloper committed:
UPDATE: Renaming Targets and joining PR-Response

gazer42 commented on README.md:
````diff
@@ -60,14 +60,11 @@ Damit das Template relativ schmerzlos vor einem Release getestet werden kann, si
-    ```zsh
-    make install-template
+    dotnet build
+    dotnet build /t:Install-Template
````
```suggestion
    msbuild -t:Install-Template
```

gazer42 commented on README.md:
```diff
@@ -82,16 +79,28 @@ Damit das Template relativ schmerzlos vor einem Release getestet werden kann, si
 
     Wenn Sie sich schon im Ordner befinden, in dem die Anwendung liegen soll, können Sie die `-o` Option durch die `-n` Option austauschen. Das verhindert, dass die Anwendung in einen Ordner gelegt wird, der zusätzlich im aktuellen Verzeichnis generiert wird.
-    Mit dem Parameter `--add-ui-tests` werden die benötigten Dateien für die UI-Tests mit angelegt.
+
+    **Optionen**:
+    | Name                 | Description                              | Type  | Default |
+    |----------------------|------------------------------------------|-------|---------|
+    | -aut, --add-ui-tests | Add UI tests to the project              | bool  | false   |
```
Hast du das manuell hinzugefügt? Weil das gibts ja schon auf develop, hättest ja einfach nur develop mergen brauchen

SeeSharpDeveloper replied:
Nee, habe den Develop gemerged, aber der hat seltsamerweise nen rebase gemacht.

gazer42 commented on README.md:
````diff
@@ -82,16 +79,28 @@ Damit das Template relativ schmerzlos vor einem Release getestet werden kann, si
 
     ```zsh
-    make run
+    dotnet build /t:Start
````
```suggestion
    msbuild -t:Start
```

SeeSharpDeveloper replied:
msbuild funktioniert nicht nativ wenn man dotnet installiert

gazer42 commented on Template.ProcessCube.Project/README.md:
````diff
@@ -1 +1,78 @@
+
+### Basic Build
+
+Wenn nur ein csproj in der Solution enthalten ist, dann kann das Projekt innerhalb der Solution gebaut werden mit
+
+```zsh
+dotnet build
+```
+
+, denn es wird dann das einzige Projekt in der Solution gebaut.
+
+Bei mehreren Projekten gehe in den Ordner apps/Template.ProcessCube.Project/
+
+```zsh
+cd apps/Template.ProcessCube.Project/
+```
````
```suggestion
```zsh
dotnet build apps/Template.ProcessCube.Project
```
```

SeeSharpDeveloper replied:
Das allein reicht auch nicht, weil man ja sonst auch noch die Targets dahinter hängen muss anschließend. Noch zu diskutieren!

gazer42 commented on Template.ProcessCube.Project/README.md:
````diff
@@ -1 +1,78 @@
+dotnet build /t:Up
+```
+
+oder wenn **Setup** und **Up** nacheinander ausgeführt werden sollen:
````
```suggestion
Wenn **Setup** und **Up** nacheinander ausgeführt werden sollen:
```

gazer42 commented on Template.ProcessCube.Project/README.md:
````diff
@@ -1 +1,78 @@
+dotnet build /t:Start
+```
+
+daraufhin wird der Docker Container zusammengebaut und gestartet.
````
```suggestion
Daraufhin wird der Docker Container zusammengebaut und gestartet.
```

gazer42 commented on TemplateHandler.csproj:
```diff
@@ -2,11 +2,11 @@
   <PropertyGroup>
     <PackageType>Template</PackageType>
-    <PackageVersion>2.1.10</PackageVersion>
+    <PackageVersion>2.2.0</PackageVersion>
```
Die Version gibt es schon. Um sowas zu vermeiden vorher einfach immer develop mergen

SeeSharpDeveloper committed:
ADD: Test Targets

SeeSharpDeveloper committed:
Add Author

SeeSharpDeveloper committed:
Merge branch 'develop' into feature/replace_makefile_with_msbuild

SeeSharpDeveloper committed:
UPDATE: Template with MsBuild and Tests
- Build Project on Setup/Start/BuildTest/Test
- Clear Test with Target "TestClean"
- Update Readme to current changes
- Suggestion to use "dotnet msbuild" for Targets

SeeSharpDeveloper committed:
Call Target "clear" after dotnet clean

SeeSharpDeveloper committed:
Delete MakefileTesting and Add Messages to Targets

gazer42 commented on Template.ProcessCube.Project/README.md:
````diff
@@ -1 +1,69 @@
+### Basic Build
+
+```zsh
+dotnet msbuild apps/Template.ProcessCube.Project
````
```suggestion
dotnet build apps/Template.ProcessCube.Project
```

Wenn es darum geht die App zu bauen würde ich nicht msbuild benutzen

gazer42 commented on Template.ProcessCube.Project/apps/Template.ProcessCube.Project/Template.ProcessCube.Project.targets:
```diff
@@ -0,0 +1,82 @@
+
+  <!-- SubTargets that can be called by Standard Targets  -->
+  <Target Name="BuildApp">
+    <MSBuild Projects="$(MSBuildProjectFile)" Targets="Build" />
```
Das ist dann quasi wie `dotnet build` richtig?

SeeSharpDeveloper replied:
Genau!

gazer42 commented on Template.ProcessCube.Project/apps/Template.ProcessCube.Project/Template.ProcessCube.Project.csproj:
```diff
@@ -28,6 +24,8 @@
   </ItemGroup>
 
+  <Import Project="$(MSBuildProjectName).targets" Condition="Exists('$(MSBuildProjectName).targets')" />
+  <Import Project="$(MSBuildProjectName).ui-tests.targets" Condition="Exists('$(MSBuildProjectName).ui-tests.targets')" />
```
Es gibt von der Template Engine aus auch eine bestimmte Syntax wie du Code-Blöcke bedingt in- oder excluden kannst. Finde das aber auch ok, könnte aber zu Inkonsistenz führen, wenn an anderen Stellen, die Syntax der Template Engine benutzt wird

SeeSharpDeveloper replied:
Das mit dem inkludieren über die Template CodeBlöcke könnte ich zusätzlich einbauen, damit nicht irgendwann ganz viele unnötige Zeilen vorhanden sind.
Das Anlegen einer eigenen Targets-Datei finde ich sinnvoll, weil dann alles von den Tests wirklich an einer Stelle abgelegt ist und die normale .targets dann nicht zu voll wird.
Mit den Inkonsitenzen müsstest du mir nochmal erklären.

gazer42 commented on Template.ProcessCube.Project/apps/Template.ProcessCube.Project/Template.ProcessCube.Project.targets:
```diff
@@ -0,0 +1,82 @@
+    <CallTarget Targets="BuildApp" />
+    <CallTarget Targets="DockerComposeUp" />
+  </Target>
+  <Target Name="Setup">
+    <Message Importance="High" Text="Starting 'Setup'" />
+    <CallTarget Targets="BuildApp" />
+  </Target>
+  <Target Name="Up">
+    <Message Importance="High" Text="Starting 'Up'" />
+    <CallTarget Targets="DockerComposeUp" />
+  </Target>
```
Weiß nicht, ob die Verschachtelung nötig ist, aber kann man so machen

gazer42 commented on Template.ProcessCube.Project/README.md:
````diff
@@ -1 +1,69 @@
+- Löscht die `node_modules` und die `package-lock.json` im Frontend
+- Löscht die Datenbank der 5Minds Engine
+
+  ```zsh
+  dotnet clean
+  ```
````
```suggestion
```zsh
dotnet clean
```
```

gazer42 commented on TemplateHandler.csproj:
```diff
@@ -2,10 +2,10 @@
   <PropertyGroup>
     <PackageType>Template</PackageType>
-    <PackageVersion>2.2.0</PackageVersion>
+    <PackageVersion>2.2.1</PackageVersion>
```
```suggestion
    <PackageVersion>2.3.0</PackageVersion>
```

Ich würde behaupten, dieser PR ist schon mehr als ein Bugfix :D

SeeSharpDeveloper replied:
Hast Recht

gazer42 commented on TemplateHandler.csproj:
```diff
@@ -2,10 +2,10 @@
     <PackageId>Template.ProcessCube.Project</PackageId>
     <Title>Project Template</Title>
-    <Authors>Jeremy Hill</Authors>
+    <Authors>Jeremy Hill, David Wehmeyer</Authors>
```
👀

SeeSharpDeveloper replied:
Kann ich auch gern wieder entfernen, musste eine Stelle für nen Commit ändern, weil der nicht das getan hat, was ich wollte. Wie wird das mit den Autoren gehandhabt?

gazer42 commented on Template.ProcessCube.Project/apps/Template.ProcessCube.Project/Template.ProcessCube.Project.ui-tests.targets:
```diff
@@ -0,0 +1,33 @@
+    <Message Importance="High" Text="Docker Compose Up 'Base + Tests'"/>
+    <Exec Command="docker compose -f docker-compose.yml -f docker-compose.tests.yml up --abort-on-container-exit" WorkingDirectory="$(DockerRoot)" ConsoleToMSBuild="true" />
+  </Target>
+</Project>
```
```suggestion
</Project>

```

gazer42 reviewed (changes requested):
Funktionalität habe ich jetzt noch nicht getestet, aber sieht erstmal gut aus soweit

SeeSharpDeveloper committed:
ADD: In "TestClean" remove Docker test container

SeeSharpDeveloper committed:
UPDATE: Opt out ui-test.targets from csproj
- Increase Version to 2.3.0
- Readme Updates

SeeSharpDeveloper committed:
ADD: Append Readme Information for UI-Tests
- realized with template conditionals

SeeSharpDeveloper committed:
ADD: Append Readme Information for UI-Tests
- realized with template conditionals

SeeSharpDeveloper committed:
ADD: Append Readme Information for UI-Tests
- realized with template conditionals

gazer42 commented on Template.ProcessCube.Project/README.md:
````diff
@@ -1 +1,121 @@
+```
+
+Bei einem folgenden `Setup` werden **alle** Dateien neu geladen und gebaut.
+---#if (add-ui-tests)
+
````
```suggestion

---#if (add-ui-tests)
```

gazer42 commented on Template.ProcessCube.Project/README.md:
```diff
@@ -1 +1,121 @@
+## UI-Test mit Robot-Framework
+
+Grundlegende UI-Tests sind in diesem Projekt bereits vorbereitet.
+Diese liegen im Ordner `./integration-tests` und benutzen das Robot Framework.
```
```suggestion
Diese liegen im Ordner `integration-tests` und benutzen das [Robot Framework](https://robotframework.org/).
```

gazer42 commented on Template.ProcessCube.Project/README.md:
````diff
@@ -1 +1,121 @@
+dotnet msbuild -t:Test
+```
+
+Jetzt sollten die UI-Tests durchlaufen und ein Result liefern.
````
```suggestion
Jetzt sollten die UI-Tests durchlaufen und ein Ergebnis liefern.
```

gazer42 commented on Template.ProcessCube.Project/README.md:
```diff
@@ -1 +1,121 @@
+
+Folgender Befehl
+
+- Löscht den zuvor erstellten Ordner mit den Testergebnissen /testresults
```
```suggestion
- Löscht den zuvor erstellten Ordner mit den Testergebnissen `testresults`
```

gazer42 commented on Template.ProcessCube.Project/apps/Template.ProcessCube.Project/Template.ProcessCube.Project.props:
```diff
@@ -0,0 +1,15 @@
+<Project>
+    <PropertyGroup>
+        <!-- SpaYarp configuration -->
+        <SpaRoot>frontend\</SpaRoot>
+        <!-- TARGET Properties-->
+        <PackageLockJsonFile>frontend\package-lock.json</PackageLockJsonFile>
+        <NodeModulesPath>frontend\node_modules</NodeModulesPath>
+        <AtlasEngineDbFile>../../.processcube/engine/db/atlasengine.sqlite</AtlasEngineDbFile>
+        <AppName>$(MSBuildProjectName)</AppName>
+        <DockerRoot>../../</DockerRoot>
+        <DockerImageName>$(MSBuildProjectName.ToLower().Replace(".","_"))</DockerImageName>
+        <SpaClientUrl>http://frontend:3000</SpaClientUrl>
```
```suggestion
        <SpaRoot>frontend\</SpaRoot>
        <SpaClientUrl>http://frontend:3000</SpaClientUrl>
        <!-- TARGET Properties-->
        <PackageLockJsonFile>frontend\package-lock.json</PackageLockJsonFile>
        <NodeModulesPath>frontend\node_modules</NodeModulesPath>
        <AtlasEngineDbFile>../../.processcube/engine/db/atlasengine.sqlite</AtlasEngineDbFile>
        <AppName>$(MSBuildProjectName)</AppName>
        <DockerRoot>../../</DockerRoot>
        <DockerImageName>$(MSBuildProjectName.ToLower().Replace(".","_"))</DockerImageName>
```

gazer42 replied:
Im Endeffekt einfach die `<SpaClientUrl>http://frontend:3000</SpaClientUrl>` nach oben schieben

gazer42 commented on Template.ProcessCube.Project/apps/Template.ProcessCube.Project/Template.ProcessCube.Project.ui-tests.targets:
```diff
@@ -0,0 +1,37 @@
+  <Target Name="TestClean">
+    <Message Importance="High" Text="Cleanup TestResults"/>
+    <RemoveDir Directories="$(TestResultsPath)" />
+     <CallTarget Targets="RemoveTestContainer"/>
```
```suggestion
    <CallTarget Targets="RemoveTestContainer"/>
```

gazer42 commented on Template.ProcessCube.Project/apps/Template.ProcessCube.Project/Template.ProcessCube.Project.ui-tests.targets:
```diff
@@ -0,0 +1,37 @@
+  <Target Name="TestUp">
+    <CallTarget Targets="DockerComposeTestUp"/>
+  </Target>
+  <Target Name="TestClean">
```
```suggestion
  <Target Name="TestClean" AfterTargets="Clean">
```

gazer42 reviewed (approved):
Würde dann noch die Makefiles löschen

SeeSharpDeveloper committed:
UPDATE: Remove makefiles and add spellfixes
- Removed makefiles, because its function is included in MsBuild-Targets
- Auto-Clean Testresults with `dotnet clean`
- spellfixes and PR suggestions

SeeSharpDeveloper committed:
Call npm install with --verbose
Use CallTarget in csproj for npmInstall and npmRunBuild
